src/script/src/python/ffi_types/vector.rs

```diff
@@ -0,0 +1,524 @@
+// limitations under the License.
+
+#[cfg(test)]
+mod tests;
+use std::ops::Deref;
+use std::sync::Arc;
+
+use datatypes::arrow::array::{
+    Array, ArrayRef, BooleanArray, Float64Array, Int64Array, UInt64Array,
+};
+use datatypes::arrow::compute;
+use datatypes::arrow::compute::kernels::{arithmetic, comparison};
+use datatypes::arrow::datatypes::DataType as ArrowDataType;
+use datatypes::arrow::error::Result as ArrowResult;
+use datatypes::data_type::DataType;
+use datatypes::prelude::Value;
+use datatypes::value::{self, OrderedFloat};
+use datatypes::vectors::{Helper, NullVector, VectorRef};
+use pyo3::pyclass as pyo3class;
+use rustpython_vm::builtins::{PyBaseExceptionRef, PyBool, PyFloat, PyInt, PyNone};
+use rustpython_vm::sliceable::{SaturatedSlice, SequenceIndex, SequenceIndexOp};
+use rustpython_vm::types::PyComparisonOp;
+use rustpython_vm::{
+    pyclass as rspyclass, AsObject, PyObject, PyObjectRef, PyPayload, PyRef, PyResult,
+    VirtualMachine,
+};
+
+use crate::python::utils::is_instance;
+
+/// The Main FFI type `PyVector` that is used both in RustPython and PyO3
+#[pyo3class(name = "vector")]
+#[rspyclass(module = false, name = "vector")]
+#[derive(PyPayload, Debug, Clone)]
+pub struct PyVector {
+    pub(crate) vector: VectorRef,
+}
+
+impl From<VectorRef> for PyVector {
+    fn from(vector: VectorRef) -> Self {
+        Self { vector }
+    }
+}
+
+fn to_type_error(vm: &'_ VirtualMachine) -> impl FnOnce(String) -> PyBaseExceptionRef + '_ {
+    |msg: String| vm.new_type_error(msg)
+}
+
+/// Performs `val - arr`.
+pub(crate) fn arrow_rsub(arr: &dyn Array, val: &dyn Array) -> Result<ArrayRef, String> {
+    arithmetic::subtract_dyn(val, arr).map_err(|e| format!("rsub error: {e}"))
+}
+
+/// Performs `val / arr`
+pub(crate) fn arrow_rtruediv(arr: &dyn Array, val: &dyn Array) -> Result<ArrayRef, String> {
+    arithmetic::divide_dyn(val, arr).map_err(|e| format!("rtruediv error: {e}"))
+}
+
+/// Performs `val / arr`, but cast to i64.
+pub(crate) fn arrow_rfloordiv(arr: &dyn Array, val: &dyn Array) -> Result<ArrayRef, String> {
+    let array =
+        arithmetic::divide_dyn(val, arr).map_err(|e| format!("rtruediv divide error: {e}"))?;
+    compute::cast(&array, &ArrowDataType::Int64).map_err(|e| format!("rtruediv cast error: {e}"))
+}
+
+pub(crate) fn wrap_result<F>(f: F) -> impl Fn(&dyn Array, &dyn Array) -> Result<ArrayRef, String>
+where
+    F: Fn(&dyn Array, &dyn Array) -> ArrowResult<ArrayRef>,
+{
+    move |left, right| f(left, right).map_err(|e| format!("arithmetic error {e}"))
+}
+
+pub(crate) fn wrap_bool_result<F>(
+    op_bool_arr: F,
+) -> impl Fn(&dyn Array, &dyn Array) -> Result<ArrayRef, String>
+where
+    F: Fn(&dyn Array, &dyn Array) -> ArrowResult<BooleanArray>,
+{
+    move |a: &dyn Array, b: &dyn Array| -> Result<ArrayRef, String> {
+        let array = op_bool_arr(a, b).map_err(|e| format!("scalar op error: {e}"))?;
+        Ok(Arc::new(array))
+    }
+}
+
+fn is_float(datatype: &ArrowDataType) -> bool {
```

Comment:
Adds `#[inline]` attributes to `is_float`, `is_signed` functions etc. These small functions should be hinted inline.